### DISCUSSION TITLE
Accessing NATS outside of Docker Container

### DISCUSSION THREAD

**Discussion by oheckmann** [nats-io/nats.go#1130] [Q&A]
The is probably an obvious mistake, but I have been banging my head on it for several days. For a project I have a Go Server running in a Docker container, a NATs Server, and a standalone client.
The server tries to send messages through NATs. To connect to NATs, it uses the following code:

    var nc, err = nats.Connect(nats.DefaultURL)
    	
    	if err != nil {
    		panic(err)
    	}
    	defer nc.Close()

But when I try to run the server, I get an error which causes the program and container to stop: `panic: nats: no servers available for connection`. This seems to mean that the program Server in the container cannot find the NATs that is running.
Both Windows-server and NATs are running in the same Docker Network.
And when I enter into the Container and use telnet to access Nats at the port nats:4222, telnet finds the NATs server correctly

~~~
/app # busybox-extras telnet nats:4222
Connected to nats:4222
INFO {"server_id":"NBFTCFLMIEXINXIRG5LSJ2MGT4J574OYHHJ7AANWDTWQERKZ76LG7RVQ","server_name":"NBFTCFLMIEXINXIRG5LSJ2MGT4J574OYHHJ7AANWDTWQERKZ76LG7RVQ","version":"2.9.6","proto":1,"git_commit":"289a9e1","go":"go1.19.3","host":"0.0.0.0","port":4222,"headers":true,"max_payload":1048576,"client_id":12,"client_ip":"172.20.0.3","cluster":"my_cluster"}
~~~
I have already tried to change the URL that ```nats.Connect``` is connecting to, and none of them have worked. I have tried ```"nats"``` and ```"nats:4222"``` and ```"nats://nats:4222"```, and the  NATs URLs as well. I do not know what I am doing incorrectly.

The Dockerfile of the Container looks like this:
~~~
FROM golang:alpine
WORKDIR /app
COPY . .
RUN apk update
RUN apk add  ffmpeg libsm libxext
RUN go mod download
RUN go build -o server
EXPOSE 50051
CMD ["./server"]
~~~

I set up NATs with the command ```docker run -d --rm --name nats  --network=mybridge -p 4222:4222 nats```, the server with ```docker run -d --name windows-server  --network=mybridge  -p 50051:50051 windows-server```
The Server works correctly when it is running outside of a container. What do I need to change so that the Go server can access Nats?

**Comment by wallyqs** [comment]
@oheckmann did you use the `docker create network` command?

**Reply by oheckmann** [reply]
Thank you for asking. Yes, I created the network with the command, such that `docker network ls` yields
`docker network ls
NETWORK ID     NAME       DRIVER    SCOPE

789db6fb27c1   bridge     bridge    local
356169b81fd7   host       host      local
11f58f38732d   kind       bridge    local
cebd069bc0ec   mybridge   bridge    local
b3115cb37054   none       null      local
44a832db80af   testnet    bridge    local
`
Is this correct? Or did I do something incorrect on this part?

**Reply by wallyqs** [reply]
You can try something like below (more or less based on docker docs example at https://docs.docker.com/network/network-tutorial-standalone/#use-user-defined-bridge-networks:

```sh
# Create network
docker network create --driver bridge nats-net

# Create nats://nats:4222 DNS entry within nats-net network and also reachablel via localhost:4222 (-p 4222:4222)
docker run -p 4222:4222 -dit --name nats --network nats-net nats:latest

# Create nats-box container in nats-net network that is aware of `nats://nats:4222` endpoint by being part of nats-net
docker run -dit --name nats-box --network nats-net natsio/nats-box
docker attach nats-box
```

Then from nats-box (which could be replaced with your app) you would be able to pubsub using the name of the nats container that got started at the beginning:

```
343caa9ba5bc:~# nc nats 4222
INFO {"server_id":"NBE6B2X3XLK7WBHZCBFCLK6VR5R3VIHEEZR63WA472ACHUAG5WXXFI47","server_name":"NBE6B2X3XLK7WBHZCBFCLK6VR5R3VIHEEZR63WA472ACHUAG5WXXFI47","version":"2.9.0","proto":1,"git_commit":"517d9b3","go":"go1.19.1","host":"0.0.0.0","port":4222,"headers":true,"max_payload":1048576,"client_id":4,"client_ip":"172.19.0.3","cluster":"my_cluster"} 
^Cpunt!

343caa9ba5bc:~# nats -s nats://nats:4222 sub foo
03:09:31 Subscribing on foo 
^C
343caa9ba5bc:~# nats -s nats://nats:4222 sub foo &
343caa9ba5bc:~# 03:09:36 Subscribing on foo 

343caa9ba5bc:~# nats -s nats://nats:4222 pub foo bar
03:09:45 Published 3 bytes to "foo"
[#1] Received on "foo"
bar
```

Try creating a nats-box sub:

```
nats -s nats://nats:4222 sub foo
28f02992f050:~# nats -s nats://nats:4222 sub foo
03:17:55 Subscribing on foo 
```

Then publish a message from outside the container:

```
nats -s localhost:4222 pub foo bar
19:18:06 Published 3 bytes to "foo"
```

You would see the message arrive on the nats-box container:

```
28f02992f050:~# nats -s nats://nats:4222 sub foo
03:17:55 Subscribing on foo 
[#1] Received on "foo"
bar
```

**Comment by wallyqs** [comment]
btw what is the environment that you are running this on? above should just work on a linux env but there might be other docker networking quirks on Windows

**Reply by oheckmann** [reply]
Thank you for the assistance, I am trying what you describe now. I am currently running it on Windows, but after I get the Docker container working, I am going to use the Docker Image in Kubernetes, on a linux system
EDIT: I followed the instructions that you suggested. When I try to attach the docker container after creating it, it just hangs forever on cli
